src/WebJobs.Script.WebHost/Helpers/ScheduledDisposer.cs

```
@@ -0,0 +1,139 @@
+        {
+            if (_semaphoreHelper.Wait(_semaphore, 0))
+            {
+                try
+                {
+                    if (_disposable == null)
+                    {
+                        _logger.LogDebug("Scheduling delayed dispose.");
+                        _disposable = disposable;
+                        ScheduleTimer();
+                        return true;
+                    }
+                    else
+                    {
+                        _logger.LogDebug("Dispose in progress.");
+                        return false;
+                    }
+                }
+                finally
+                {
+                    _semaphoreHelper.Release(_semaphore);
+                }
+            }
+
+            _logger.LogDebug("Failed to acquire lock.");
+            return false;
+        }
+
+        public bool ScheduleDispose(IDisposable disposable)
+        {
+            var disposeScheduled = false;
+
+            try
+            {
+                disposeScheduled = TryScheduleDispose(disposable);
+            }
+            catch (Exception e)
+            {
+                // fallback to disposing on current thread
+                _logger.LogError(e, $"Failed to {nameof(TryScheduleDispose)}");
+            }
+
+            if (!disposeScheduled)
+            {
+                _logger.LogDebug($"Disposing {nameof(disposable)} on current thread.");
+                disposable.Dispose();
+            }
+
+            return disposeScheduled;
+        }
+
+        private void ScheduleTimer()
+        {
+            var timer = _timer;
+            try
+            {
+                timer?.Change(_delayTimeMs, Timeout.Infinite);
+            }
+            catch (ObjectDisposedException)
+            {
+                // might race with dispose
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, $"{nameof(ScheduleTimer)}");
```

Let's not swallow this exception here. This might well be something we can't recover from.

---

Good point. Will fix this.

---

removed this code